src/mlpack/methods/ann/layer/convolution_impl.hpp

```diff
@@ -253,41 +239,47 @@ void Convolution<
     InputDataType,
     OutputDataType
 >::Backward(
-    const arma::Mat<eT>& /* input */, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
+    const arma::Mat<eT>& input, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
 {
   arma::cube mappedError(((arma::Mat<eT>&) gy).memptr(), outputWidth,
       outputHeight, outSize * batchSize, false, false);
+  arma::cube inputTemp(((arma::Mat<eT>&) input).memptr(), inputWidth,
+      inputHeight, inSize * batchSize, false, false);
 
-  g.set_size(inputWidth * inputHeight * inSize, batchSize);
-  gTemp = arma::Cube<eT>(g.memptr(), inputWidth, inputHeight,
-      inSize * batchSize, false, false);
+  g.set_size(inputTemp.n_rows * inputTemp.n_cols * inSize, batchSize);
+  gTemp = arma::Cube<eT>(g.memptr(), inputTemp.n_rows, inputTemp.n_cols,
+                         inputTemp.n_slices, false, false);
   gTemp.zeros();
-
-  for (size_t outMap = 0, outMapIdx = 0, batchCount = 0; outMap <
-      outSize * batchSize; outMap++)
+  #pragma omp parallel
+  for (size_t outMapIdx = 0; outMapIdx < outSize; outMapIdx++)
   {
-    if (outMap != 0 && outMap % outSize == 0)
-    {
-      batchCount++;
-      outMapIdx = 0;
-    }
-
-    for (size_t inMap = 0; inMap < inSize; inMap++, outMapIdx++)
+    for (size_t inMap = 0; inMap < inSize; inMap++)
     {
-      arma::Mat<eT> output, rotatedFilter;
-      Rotate180(weight.slice(outMapIdx), rotatedFilter);
-
-      BackwardConvolutionRule::Convolution(mappedError.slice(outMap),
-          rotatedFilter, output, strideWidth, strideHeight);
-
-      if (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0)
-      {
-        gTemp.slice(inMap + batchCount * inSize) += output.submat(padWLeft,
-            padHTop, padWLeft + gTemp.n_rows - 1, padHTop + gTemp.n_cols - 1);
-      }
-      else
-      {
-        gTemp.slice(inMap + batchCount * inSize) += output;
+      arma::Mat<eT> rotatedFilter;
+      Rotate180(weight.slice(outMapIdx+inMap), rotatedFilter);
+      #pragma omp for
+      for (omp_size_t batchCount = 0; batchCount < batchSize; batchCount++) {
```

Comment:
Sorry, I didn't look in detail yet. Why did you change the nesting of the loops? I was wondering if it provides any cache optimizations.

Reply:
Yes, we avoid extra computations. Earlier, we were retrieving and rotating the same weight slice for every batch, but now we only do it once for all batches.